waf/internal/bindings/waf.go

```diff
@@ -194,16 +187,146 @@ func valueToWAFInput(v reflect.Value) (in *C.PWArgs, err error) {
 	case reflect.Uint32:
 		fallthrough
 	case reflect.Uint64:
-		arg := C.powerwaf_createUint((C.uint64_t)(v.Uint()))
-		return &arg, nil
+		return makeWAFUInt((*WAFUInt)(v), data.Uint())
 	}
 }
 
-//export goOnLogMessage
-func goOnLogMessage(level C.PW_LOG_LEVEL, _, _ *C.char, _ C.int, message *C.char, length C.size_t) {
-	fmt.Println(C.GoStringN(message, C.int(length)))
+func marshalWAFMap(data reflect.Value, v *WAFMap, depth int) error {
+	// Only allow string key types
+	if data.Type().Key().Kind() != reflect.String {
+		return errors.Errorf("unexpected WAF map key type `%T` instead of `string`", data.Interface())
+	}
+	// Marshal map entries
+	for i, iter := 0, data.MapRange(); iter.Next(); i++ {
+		entry := v.Index(i)
+		// Add the key first in order to get key insertion errors before traversing
+		// the value. It would be a waste if in the end the key cannot be added.
+		key := iter.Key().String()
+		if err := makeWAFMapKey(entry, key); err != nil {
+			return errors.Wrap(err, "could not add a new map key")
+		}
+		// Marshal the key's value
+		if err := marshalWAFValueRec(iter.Value(), (*WAFValue)(entry), depth); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func marshalWAFArray(data reflect.Value, v *WAFArray, depth int) error {
+	// Profiling shows `data.Len()` is called every loop if it is
+	// used in the loop condition.
+	l := data.Len()
+	for i := 0; i < l; i++ {
+		if err := marshalWAFValueRec(data.Index(i), v.Index(i), depth); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func makeWAFMap(v *WAFMap, len int) error {
+	return makeWAFLengthedValue((*WAFValue)(v), len, C.PWI_MAP)
 }
 
-func SetupLogging() {
-	C.powerwaf_setupLogging(C.powerwaf_logging_cb_t(C.onLogMessage), C.PWL_DEBUG)
+func (m *WAFMap) Index(i int) *WAFMapEntry {
+	entry := (*WAFArray)(m).Index(i)
+	return (*WAFMapEntry)(entry)
+}
+
+func (a *WAFArray) Index(i int) *WAFValue {
+	if C.uint64_t(i) >= a.nbEntries {
+		panic(errors.New("out of bounds access to WAFArray"))
```

Comment:
Hmm, really not keen on those panics. Looks like you're beginning to use them a lot like exceptions instead of errors

Reply:
Only because `reflect` uses them this way too and that binding accessors are only based on `reflect`. So I am using the same principle in case of unexpected behavior.

For example:
```
reflect.ValueOf([]int{1, 2, 3}).Index(33) // panic: out of bound access
```

So I panic too for this specific case.